src/core/src/descriptor/shared_tensor.cpp

```diff
@@ -4,135 +4,302 @@
+            });
+        }
     }
 
-    std::shared_ptr<ITensorDescriptor> m_shared_tensor;
-    std::unordered_set<std::string> m_output_names;
-    std::unordered_set<std::string>::const_iterator m_name_it;
+    const std::weak_ptr<Tensor> m_descriptor_owner;  // tensor which holds this descriptor implementation
+    TensorNames m_assigned_names;
+    std::weak_ptr<InputTensorWrapper> m_self_observer;  // observe this class shared pointer instance
+};
+
+class ParameterTensorWrapper : public InputTensorWrapper {
+public:
+    explicit ParameterTensorWrapper(ITensorDescriptorPtr descriptor, std::weak_ptr<Tensor> descriptor_owner)
+        : InputTensorWrapper{std::move(descriptor), std::move(descriptor_owner)} {}
+
+    // --- ParameterTensorWrapper specific interface
+    static auto make(ITensorDescriptorPtr&& descriptor, std::weak_ptr<Tensor>&& descriptor_owner) {
+        auto input_tensor =
+            std::make_shared<ParameterTensorWrapper>(std::move(descriptor), std::move(descriptor_owner));
+        input_tensor->set_self_observer(input_tensor);
+        return input_tensor;
+    }
 };
 
 /**
- * @brief Set output tensor descriptor with shared tensor from new input.
+ * @brief Dedicated tensor descriptor implementation to share input descriptor.
+ *
+ * Shared tensor share input tensor but have specific properties:
+ * - tensor names - if set these are used as descriptor names and appended to input tensor because is same tensor
+ */
+class SharedTensor : public TensorWrapper {
+public:
+    SharedTensor(std::shared_ptr<InputTensorWrapper> tensor)
+        : TensorWrapper{tensor->base()},
+          m_input_descriptor{std::move(tensor)},
+          m_output_names{},
+          m_name_it{} {}
+
+    // --- ITensorDescriptor API
+    void set_names(const TensorNames& names) override {
+        rm_names(m_output_names);
+        m_output_names = names;
+        m_name_it = std::min_element(m_output_names.begin(), m_output_names.end());
+        TensorWrapper::add_names(m_output_names);
+    }
+
+    void add_names(const TensorNames& names) override {
+        m_output_names.insert(names.begin(), names.end());
+        m_name_it = std::min_element(m_output_names.begin(), m_output_names.end());
+        TensorWrapper::add_names(names);
+    }
+
+    const TensorNames& get_names() const override {
+        return m_output_names.empty() ? TensorWrapper::get_names() : m_output_names;
+    }
+
+    const std::string& get_any_name() const override {
+        return m_output_names.empty() ? TensorWrapper::get_any_name() : *m_name_it;
+    }
+
+    // --- SharedTensor specific interface
+    void set_tensor(std::shared_ptr<InputTensorWrapper> tensor) {
+        m_input_descriptor = std::move(tensor);
+        set_tensor(m_input_descriptor->base());
+    }
+
+    auto get_input_descriptor() const {
+        return m_input_descriptor;
+    }
+
+    const auto& get_assigned_names() const {
```

Comment:
```suggestion
    const auto& get_assigned_names() & {
```
to be sure nobody returns internal state of temporary obj

Reply:
Could you explain more?
The getter is similiar like `get_names`.
Why it should be different?